### PR TITLE
Fix #3854: Add Either.cond like constructor to IO, RIO, UIO, URIO, ZIO

### LINKED ISSUE #3854: Cond like constructor
Hello I was looking for a method like cond on Either. It would work like so:

```
object ZIO {
 def cond[E, A](predicate: Boolean, error: => E, result: => A): IO[E, A] = if (predicate) ZIO.succeed(result) else ZIO.fail(error)
}
```

is there another way of doing this? if not would you consider a pr with a method like this?

### PR DESCRIPTION
**Pull request by gerdreiss**
I'm not sure whether it makes sense to add the constructor to UIO and URIO, since these types do not presume errors.

### CODE REVIEW

**Comment by adamgfraser** [comment]
Yes you're 100% right. It should only be on types that can fail and for types where the error type is fixed to a particular type such as `Throwable` the constructor should be fixed to that error type as well.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -2401,6 +2401,13 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   def collectParN[R, E, A, B](n: Int)(in: Iterable[A])(f: A => ZIO[R, Option[E], B]): ZIO[R, E, List[B]] =
     foreachParN(n)(in)(a => f(a).optional).map(_.flatten)
 
+  /**
+   * Similar to Either.cond, evaluate the predicate,
+   * return the given A as success if predicate returns true, and the given E as error otherwise
+   */
+  def cond[R, E, A](predicate: Boolean, result: => A, error: => E): ZIO[R, E, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala:2408]:
```suggestion
  def cond[E, A](predicate: Boolean, result: => A, error: => E): IO[E, A] =
```

File: core/shared/src/main/scala/zio/RIO.scala [hunk 2]
```diff
@@ -206,6 +206,12 @@ object RIO {
   def collectParN[R, E, A, B](n: Int)(in: Iterable[A])(f: A => ZIO[R, Option[Throwable], B]): RIO[R, List[B]] =
     ZIO.collectParN(n)(in)(f)
 
+  /**
+   * @see See [[zio.ZIO.cond]]
+   */
+  def cond[R, A](predicate: Boolean, result: => A, error: => Throwable): RIO[R, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/RIO.scala:212]:
```suggestion
  def cond[A](predicate: Boolean, result: => A, error: => Throwable): Task[A] =
```

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 1 above)
Comment (gerdreiss) [reply core/shared/src/main/scala/zio/ZIO.scala:2408]: True, the 'R' is totally useless in my version

File: core/shared/src/main/scala/zio/RIO.scala (hunk 2 above)
Comment (gerdreiss) [reply core/shared/src/main/scala/zio/RIO.scala:212]: Agreed

**Comment by gerdreiss** [comment]
I've added constructors to ZManaged and ZSTM you've mentioned in my previous PR. Not 100% sure, all is correct, though it does seem to make sense...

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 3]
```diff
@@ -1463,6 +1463,13 @@ object ZManaged extends ZManagedPlatformSpecific {
   def collectParN[R, E, A, B](n: Int)(in: Iterable[A])(f: A => ZManaged[R, Option[E], B]): ZManaged[R, E, List[B]] =
     foreachParN(n)(in)(a => f(a).optional).map(_.flatten)
 
+  /**
+   * Similar to Either.cond, evaluate the predicate,
+   * return the given A as success if predicate returns true, and the given E as error otherwise
+   */
+  def cond[R, E, A](predicate: Boolean, result: => A, error: => E): ZManaged[R, E, A] =
+    if (predicate) effectTotal(result) else fail(error)
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZManaged.scala:1471]: Can use `succeed` here.

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 4]
```diff
@@ -1034,6 +1034,13 @@ object ZSTM {
   def collectAll_[R, E, A](in: Chunk[ZSTM[R, E, A]]): ZSTM[R, E, Unit] =
     foreach_(in)(ZIO.identityFn)
 
+  /**
+   * Similar to Either.cond, evaluate the predicate,
+   * return the given A as success if predicate returns true, and the given E as error otherwise
+   */
+  def cond[R, E, A](predicate: Boolean, result: => A, error: => E): ZSTM[R, E, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1041]: Don't need `R` parameter. Can also add alias to `STM` object.

Comment (gerdreiss) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1041]: Should I add one to TaskSTM, similar to RIO?

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 5]
```diff
@@ -1463,6 +1463,13 @@ object ZManaged extends ZManagedPlatformSpecific {
   def collectParN[R, E, A, B](n: Int)(in: Iterable[A])(f: A => ZManaged[R, Option[E], B]): ZManaged[R, E, List[B]] =
     foreachParN(n)(in)(a => f(a).optional).map(_.flatten)
 
+  /**
+   * Similar to Either.cond, evaluate the predicate,
+   * return the given A as success if predicate returns true, and the given E as error otherwise
+   */
+  def cond[R, E, A](predicate: Boolean, result: => A, error: => E): ZManaged[R, E, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZManaged.scala:1470]: No `R` parameter here.

Comment (gerdreiss) [reply core/shared/src/main/scala/zio/ZManaged.scala:1470]: Oh, of course.

**Comment by gerdreiss** [comment]
Well, with so much help it could only get perfect! ;-))

**Pull request merged**